### PR TITLE
Added pressure handler to options

### PR DESCRIPTION
**Pull request by Jelenkee**
Heyho,
I used this plugin in one of my projects and I don't know why there is no way to just add a handler to the options to handle the 'pressure'. There was no way to determine what kind of pressure happend (eventLoopDelay, heapBytes, ...) and you could not handle the pressure and proceed the request the usual way.
So I added a way to add a pressure handler to the options.
 
#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

**Comment by mcollina** [comment]
The feature is good but CI is failing badly :(

File: index.js [hunk 1]
```diff
@@ -229,3 +245,5 @@ module.exports = fp(underPressure, {
   fastify: '3.x',
   name: 'under-pressure'
 })
+
+Object.assign(module.exports, { TYPE_EVENT_LOOP_DELAY, TYPE_EVENT_LOOP_UTILIZATION, TYPE_HEALTH_CHECK, TYPE_HEAP_USED_BYTES, TYPE_RSS_BYTES })
```
Comment (mcollina) [inline index.js:249]: Please use  explicit `module.exports.SOMETHING =` statements, so that they can be used with native esm.

File: index.js [hunk 2]
```diff
@@ -155,36 +162,45 @@ async function underPressure (fastify, opts) {
...
   }
 
-  function sendError (reply, next) {
-    reply.status(SERVICE_UNAVAILABLE).header('Retry-After', retryAfter)
-    next(underPressureError)
+  function handlePressure (req, reply, next, type, value) {
+    if (typeof pressureHandler === 'function') {
+      const result = pressureHandler(req, reply, type, value)
+      if (result instanceof Promise) {
+        result.then(() => next()).catch(next)
```
Comment (mcollina) [inline index.js:196]:
```suggestion
        result.then(() => next()), next)
```

File: test.js [hunk 3]
```diff
@@ -730,6 +730,215 @@ test('Custom health check', t => {
...
+
+    fastify.register(underPressure, {
+      healthCheck: async () => false,
+      healthCheckInterval: 1,
+      pressureHandler: (req, rep, type, value) => {}
+    })
+
+    fastify.get('/', (req, rep) => rep.send('A'))
+
+    fastify.inject().get('/').end().then(res => t.equal(res.body, 'A'))
```
Comment (mcollina) [inline test.js:808]: please use async/await instead

File: test.js [hunk 4]
```diff
@@ -730,6 +730,215 @@ test('Custom health check', t => {
...
+    const errorMessage = 'promiseError'
+
+    fastify.register(underPressure, {
+      healthCheck: async () => false,
+      healthCheckInterval: 1,
+      pressureHandler: (req, rep, type, value) => new Promise((resolve, reject) => {
+        setTimeout(() => {
+          reject(new Error(errorMessage))
+        }, 250)
+      })
```
Comment (mcollina) [inline test.js:788]: please use async/await and

```js
const  { promisify } = require('util')
const sleep = promisify(setTimeout)
```

File: index.js [hunk 5]
```diff
@@ -155,36 +162,45 @@ async function underPressure (fastify, opts) {
...
+      } else {
```
Comment (mcollina) [inline index.js:197]: I think you should handle the case where result is something else of a promise.

**Review by mcollina — CHANGES_REQUESTED** [review]

File: index.js (hunk 5 above)
Comment (Jelenkee) [reply index.js:197]: But how? What am I supposed to do with a non-Promise? Calling `reply.send()` with the object/string/etc?

**Comment by mcollina** [comment]
CI is falling on every environment, there is a typo.

**Comment by Jelenkee** [comment]
> CI is falling on every environment, there is a typo.

Guess why
image
One bracket too many 😉

File: index.js (hunk 1 above)
Comment (Jelenkee) [reply index.js]: done

File: test.js (hunk 3 above)
Comment (Jelenkee) [reply test.js]: done

File: test.js (hunk 4 above)
Comment (Jelenkee) [reply test.js]: done

**Comment by Jelenkee** [comment]
These test are quite tricky. Sometimes they fail and sometimes they pass even if I didn't change the code. Even local.
7/9 is the best I can provide atm.  
`event loop delay` is failing on both systems though I copied the usual maxEventLoopDelay test and added the pressure handler.

File: index.js (hunk 5 above)
Comment (mcollina) [reply index.js:197]: Either that or error the response

Comment (Jelenkee) [reply index.js]: done

**Comment by Jelenkee** [comment]
Again 7/9. But this times other environments failed

File: test.js [hunk 6]
```diff
@@ -730,6 +730,230 @@ test('Custom health check', t => {
...
+      }
+    })
+
+    fastify.get('/', (req, rep) => rep.send('A'))
+
+    fastify.listen(0, async (err, address) => {
+      t.error(err)
+      fastify.server.unref()
+
+      if (monitorEventLoopDelay) {
```
Comment (mcollina) [inline test.js:844]: I think if this is not available we should skip the test.

File: test.js [hunk 7]
```diff
@@ -730,6 +730,230 @@ test('Custom health check', t => {
...
+        await wait(500)
+      }
+      process.nextTick(() => block(1000))
```
Comment (mcollina) [inline test.js:847]:
```suggestion
      process.nextTick(() => block(1500))
```

**Review by mcollina — CHANGES_REQUESTED** [review]

File: test.js (hunk 6 above)
Comment (Jelenkee) [reply test.js]: done but it didnt fix the failing tests completely

**Comment by mcollina** [comment]
The tests fails for me locally as well on Mac, it's not a flake.

**Comment by mcollina** [comment]
I recommend to split test.js into two (or more) files under the test/ folder. I suspect there might be a few problems of concurrency/interaction between the tests. You might also want to rebase on top of master, as well as adding disabling parallelism of tests in tap.

**Review by mcollina — APPROVED** [review]
lgtm

**Comment by Jelenkee** [comment]
Finally all tests are green.  
Quick! Merge as soon as possible before they turn red again 😉

**Comment by mcollina** [comment]
Deal!

Thanks for bearing with me. This module is _really tricky_.

**Pull request merged**